File: GanAndroidApp/app/src/main/java/com/xattacker/android/main/MainActivity.java
package com.xattacker.android.main;

import android.app.Activity;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.os.Looper;
import android.provider.Settings;
import android.view.Gravity;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.xattacker.gan.GanClient;
import com.xattacker.gan.GanClientListener;
import com.xattacker.util.DateTimeUtility;
import com.xattacker.util.DateTimeUtility.DateTimeFormatType;

import java.util.Date;

public class MainActivity extends Activity implements OnClickListener, GanClientListener
{
	 private EditText _accountEdit, _pwdEdit;
	 private EditText _receiverEdit, _msgEdit;
	 private EditText _receivedMsgEdit;
	
    @Override
    public void onCreate(Bundle savedInstanceState) 
    {
        super.onCreate(savedInstanceState);
        
        try
        {
      	  // 122.254.39.227 // home
      	  // 192.168.226.50 // office
      	  GanClient.constructInstance("192.168.226.50", 5999, this);
        }
        catch (Exception ex)
        {
      	  android.util.Log.i("aaa", ex.toString());
        }
        
        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);
        setContentView(layout);
        
        
        _accountEdit = new EditText(this);
        _accountEdit.setText("test");
        layout.addView(_accountEdit);
        
        _pwdEdit = new EditText(this);
        _pwdEdit.setText("123");
        layout.addView(_pwdEdit);
        
        LinearLayout layout2 = new LinearLayout(this);
        layout2.setOrientation(LinearLayout.HORIZONTAL);
        layout.addView(layout2);
        
        Button button = new Button(this);
        button.setText("login");
        button.setId(1);
        button.setOnClickListener(this);
        layout2.addView(button);
        
        button = new Button(this);
        button.setText("logout");
        button.setId(2);
        button.setOnClickListener(this);
        layout2.addView(button);
        
        button = new Button(this);
        button.setText("get ip");
        button.setId(4);
        button.setOnClickListener(this);
        layout2.addView(button);

        button = new Button(this);
        button.setText("get server time");
        button.setId(5);
        button.setOnClickListener(this);
        layout2.addView(button);
        
        
        _receiverEdit = new EditText(this);
        _receiverEdit.setText("test");
        layout.addView(_receiverEdit);
        
        _msgEdit = new EditText(this);
        _msgEdit.setText("1231234");
        layout.addView(_msgEdit);
        
        button = new Button(this);
        button.setText("send sms");
        button.setId(3);
        button.setOnClickListener(this);
        layout.addView(button);
        
        
        _receivedMsgEdit = new EditText(this);
        _receivedMsgEdit.setGravity(Gravity.TOP);
        _receivedMsgEdit.setTextColor(Color.BLUE);
        _receivedMsgEdit.setEnabled(false);
        _receivedMsgEdit.setLines(3);
        layout.addView(_receivedMsgEdit);
    }
    
    public void onDestroy()
    {
		 super.onDestroy();

		 GanClient.release();
    }

	@Override
	public void onClick(View aView)
	{
		switch (aView.getId())
		{
			case 1:
			{
				final String account = _accountEdit.getText().toString();
				final String pwd = _pwdEdit.getText().toString();

				asyncRun(
				new AsyncRunnable()
				{
					@Override
					public void run() throws Exception
					{
						boolean result = GanClient.instance().accountService().login(account, pwd);
						android.util.Log.i("aaa", "login " + result);
						showToast("login result: " + result);
					}

					@Override
					public void onExceptionHappened(Throwable th)
					{
					}
				}
				);
			}
				break;
				
			case 2:
			{
				asyncRun(
				new AsyncRunnable()
				{
					@Override
					public void run() throws Exception
					{
						boolean result = GanClient.instance().accountService().logout();
						android.util.Log.i("aaa", "logout " + result);
						showToast("logout result: " + result);
					}

					@Override
					public void onExceptionHappened(Throwable th)
					{
					}
				}
				);

				_receivedMsgEdit.setText("");
			}
				break;
				
			case 3:
			{
				final String receiver = _receiverEdit.getText().toString();
				final String msg = _msgEdit.getText().toString();

				asyncRun(
				new AsyncRunnable()
				{
					@Override
					public void run() throws Exception
					{
						boolean result = GanClient.instance().smsService().sendMsg(receiver, msg);
						android.util.Log.i("aaa", "send msg " + result);
						showToast("send msg result: " + result);
					}

					@Override
					public void onExceptionHappened(Throwable th)
					{
					}
				}
				);
			}
				break;
				
			case 4:
			{
				asyncRun(
				new AsyncRunnable()
				{
					@Override
					public void run() throws Exception
					{
						String ip = GanClient.instance().systemService().getIPAddress();
						if (ip != null)
						{
							android.util.Log.i("aaa", "get IP: " + ip);
							showToast("get IP:  " + ip);
						}
						else
						{
							showToast("get IP failed ");
						}
					}

					@Override
					public void onExceptionHappened(Throwable th)
					{
					}
				}
				);
			}
				break;
				
			case 5:
			{
				Date time = GanClient.instance().systemService().getSystemTime();
				if (time != null)
				{
					String time_str = DateTimeUtility.getDateTimeString
					                  (
					                  time, 
					                  DateTimeFormatType.DATETIME_COMPLETE
					                  );
					Toast.makeText(this, time_str, Toast.LENGTH_LONG).show();
				}

				asyncRun(
				new AsyncRunnable()
				{
					@Override
					public void run() throws Exception
					{
						Date time = GanClient.instance().systemService().getSystemTime();
						if (time != null)
						{
							String time_str = DateTimeUtility.getDateTimeString
							(
							time,
							DateTimeFormatType.DATETIME_COMPLETE
							);
							android.util.Log.i("aaa", "get time: " + time_str);
							showToast("get time:  " + time_str);
						}
						else
						{
							showToast("get time failed");
						}
					}

					@Override
					public void onExceptionHappened(Throwable th)
					{
					}
				}
				);
			}
				break;
		}
	}
	
	@Override
	public void onAccountLogined(String aAccount)
	{
		android.util.Log.i("aaa", "onAccountLogined " + aAccount);
	}
	
	@Override
	public void onAccountClosed(String aAccount)
	{
		android.util.Log.i("aaa", "onAccountClosed " + aAccount);
	}
	
	@Override
	public void onSMSReceived(final String aSender, final long aTime, final String aMsg)
	{
		android.util.Log.i("aaa", "onSMSReceived " + aSender + ", " + aTime);
		
		runOnUiThread
		(
		new Runnable()
		{
			public void run()
			{
				if (_receivedMsgEdit != null)
				{
					_receivedMsgEdit.setText("");

					Date date = new Date(aTime);
					_receivedMsgEdit.append
					(
				   DateTimeUtility.getDateTimeString
				   (
				   date, 
				   DateTimeFormatType.DATETIME_COMPLETE
				   )
				   );
					_receivedMsgEdit.append("\n" + aMsg);
				}
			}
		}
		);
	}
	public interface AsyncRunnable
	{
		void run() throws Exception;
		void onExceptionHappened(Throwable th);
	}

	protected void asyncRun(final AsyncRunnable run, final int delay)
	{
		try
		{
			new Thread()
			{
				public void run()
				{
					try
					{
						Looper.prepare();

						if (delay > 0)
						{
							Thread.sleep(delay);
						}

						run.run();
					}
					catch (Throwable th)
					{
						showToast(th.getLocalizedMessage() != null ? th.getLocalizedMessage()  : th.toString());
						run.onExceptionHappened(th);
					}
				}
			}.start();
		}
		catch (Throwable th)
		{
			showToast(th.getLocalizedMessage());
			run.onExceptionHappened(th);
		}
	}

	protected void asyncRun(final AsyncRunnable run)
	{
		asyncRun(run, 0);
	}

	public String getAppVersion()
	{
		String version = null;

		try
		{
			version = getPackageManager().getPackageInfo(getPackageName(), 0).versionName;
		}
		catch (Exception ex)
		{
		}

		return version;
	}

	protected void showToast(final String text)
	{
		if (isMainThread())
		{
			Toast.makeText(this, text, Toast.LENGTH_LONG).show();
		}
		else
		{
			runOnUiThread(new Runnable() {
				@Override
				public void run() {
					Toast.makeText(MainActivity.this, text, Toast.LENGTH_LONG).show();
				}
			});
		}
	}

	protected boolean isMainThread()
	{
		return Looper.getMainLooper().equals(Looper.myLooper());
	}

	protected boolean isOnEmulator()
	{
		String android_id = Settings.Secure.ANDROID_ID;
		String product = Build.PRODUCT;

		return "sdk".equalsIgnoreCase(product) ||
					"google_sdk".equalsIgnoreCase(product) ||
					"sdk_x86".equalsIgnoreCase(product) ||
					"sdk_google_phone_x86".equalsIgnoreCase(product) ||
					"vbox86p".equalsIgnoreCase(product) ||
					android_id == null ||
					android_id.equalsIgnoreCase("9774D56D682E549C");
	}
}

File: GanAndroidApp/app/src/main/java/com/xattacker/android/main/MainActivity.kt
package com.xattacker.android.main

import android.app.Activity
import android.os.Build
import android.os.Bundle
import android.os.Looper
import android.provider.Settings
import android.util.Log
import android.view.View
import android.widget.Toast

import com.xattacker.android.main.databinding.ActivityMainBinding
import com.xattacker.gan.GanClient
import com.xattacker.gan.GanClientListener
import com.xattacker.gan.data.MessageData
import com.xattacker.util.DateTimeFormatType
import com.xattacker.util.DateTimeUtility

import java.util.*

class MainActivity() : Activity(), GanClientListener
{
    companion object
    {
        private val ACCOUNT = "test2"
        private val PASSWORD = "test2"
    }

    private lateinit var binding: ActivityMainBinding

    public override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)

        // use view Binding mode
        binding = ActivityMainBinding.inflate(this.layoutInflater)
        setContentView(binding.root)

        this.title = "GanClientApp"

        this.loadIP("GanClient", {
            ip: String, port: Int ->
            binding.editIp.setText(ip)
            binding.editPort.setText(port.toString())
        })

        binding.editReceiver.setText(PASSWORD)
    }

    public override fun onDestroy()
    {
        super.onDestroy()

        GanClient.release()
    }

    public fun onLoginClick(view: View)
    {
        val account = ACCOUNT
        val pwd = PASSWORD

        asyncRun(
            {
                val result: Boolean = GanClient.instance?.accountService?.login(account, pwd) ?: false
                Log.i("aaa", "login $result")
                showToast("login result: $result")
            })
    }

    public fun onLogoutClick(view: View)
    {
        asyncRun(
            {
                val result: Boolean = GanClient.instance?.accountService?.logout() ?: false
                Log.i("aaa", "logout $result")
                showToast("logout result: $result")
            })

        binding.textGotMessage.setText("")
    }

    public fun onGetIPClick(view: View)
    {
        asyncRun(
            {
                val ip = GanClient.instance?.systemService?.getIP()
                if (ip != null)
                {
                    Log.i("aaa", "get IP: $ip")
                    showToast("get IP:  $ip")
                }
                else
                {
                    showToast("get IP failed ")
                }
            })
    }

    public fun onGetTimeClick(view: View)
    {
        asyncRun(
            {
                val time = GanClient.instance?.systemService?.getSystemTime()
                if (time != null)
                {
                    val time_str: String = DateTimeUtility.getDateTimeString(time, DateTimeFormatType.DATETIME_COMPLETE)
                    Log.i("aaa", "get time: $time_str")
                    showToast("get time:  $time_str")
                }
                else
                {
                    showToast("get time failed")
                }
            })
    }

    public fun onSendMsgClick(view: View)
    {
        val receiver = binding.editReceiver.text.toString()
        val msg = "send test message=" + System.currentTimeMillis()

        asyncRun(
            {
                val result: Boolean = GanClient.instance?.messageService?.sendMessage(receiver, msg) ?: false
                Log.i("aaa", "send msg $result")
                showToast("send msg result: $result")
            })
    }

    override fun onAccountLoggedIn(account: String)
    {
        Log.i("aaa", "onAccountLoggedIn $account")

        this.title = account + "(loggedIn)"
    }

    override fun onAccountLoggedOut(account: String)
    {
        Log.i("aaa", "onAccountLoggedOut $account")

        this.title = "GanClientApp"
    }

    override fun onMessageReceived(message: MessageData)
    {
        Log.i("aaa", "onMessageReceived $message")

        runOnUiThread(Runnable {

            binding.textGotMessage.setText("")

            val date = Date(message.time ?: 0)
            binding.textGotMessage.append(DateTimeUtility.getDateTimeString(date, DateTimeFormatType.DATETIME_COMPLETE))
            binding.textGotMessage.append("\n" + message.message)
        })
    }

    private fun asyncRun(run: () -> Unit, onError: ((th: Throwable) -> Unit)? = null, delay: Int = 0)
    {
        try
        {
            if (GanClient.instance == null)
            {
                try
                {
                    val ip = binding.editIp.text.toString()
                    val port = binding.editPort.text.toString().toInt()
                    GanClient.initial(ip, port, this)

                    saveIP("GanClient", ip, port)
                }
                catch (th: Throwable)
                {
                    showToast("initial GanClient failed: " + th.toString())
                    onError?.invoke(th)

                    return
                }
            }


            object : Thread()
            {
                override fun run()
                {
                    try
                    {
                        Looper.prepare()
                        if (delay > 0)
                        {
                            sleep(delay.toLong())
                        }
                        run()
                    }
                    catch (th: Throwable)
                    {
                        showToast(if (th.localizedMessage != null) th.localizedMessage else th.toString())
                        onError?.invoke(th)
                    }
                }
            }.start()
        }
        catch (th: Throwable)
        {
            showToast(th.localizedMessage)
            onError?.invoke(th)
        }
    }

    private fun showToast(text: String)
    {
        if (isMainThread)
        {
            Toast.makeText(this, text, Toast.LENGTH_LONG).show()
        }
        else
        {
            runOnUiThread(object : Runnable
            {
                override fun run()
                {
                    Toast.makeText(this@MainActivity, text, Toast.LENGTH_LONG).show()
                }
            })
        }
    }

    private val isMainThread: Boolean
         get() = (Looper.getMainLooper() == Looper.myLooper())

    private val isOnEmulator: Boolean
        get()
        {
            val android_id: String = Settings.Secure.ANDROID_ID
            val product = Build.PRODUCT
            return ("sdk".equals(product, ignoreCase = true) ||
                    "google_sdk".equals(product, ignoreCase = true) ||
                    "sdk_x86".equals(product, ignoreCase = true) ||
                    "sdk_google_phone_x86".equals(product, ignoreCase = true) ||
                    "vbox86p".equals(product, ignoreCase = true) ||
                    (android_id == null) ||
                    android_id.equals("9774D56D682E549C", ignoreCase = true))
        }

    private fun loadIP(aName: String, loaded: (ip: String, port: Int) -> Unit)
    {
        val pref = getSharedPreferences(aName, MODE_PRIVATE)
        val ip = pref.getString("IP", null)
        val port = pref.getInt("PORT", 0)
        if (ip != null && port != 0)
        {
            loaded(ip, port)
        }
    }

    private fun saveIP(name: String, ip: String, port: Int)
    {
        val pref = getSharedPreferences(name, MODE_PRIVATE)
        pref.edit().putString("IP", ip).putInt("PORT", port).commit()
    }
}